### PR TITLE
Upgrade to Jline3

### PR DESCRIPTION
**Pull request by dimitarp**

### CODE REVIEW

**Comment by ChristianGruen** [comment]
Dimitar, thanks again for your pull request.

Regarding JLine 3, I’m still a bit reluctant to switch to the new version, as…

* The format of the history file has changed (the one that’s stored in `.basexhistory`). We could probably tackle that by choosing another file name.
* The size of the library files has doubled (well, it’s still in the range of kilobytes).
* We might need to invest some time to test it with different platforms.

When I run your commits with Windows 10, I currently get the following output (what do you get?):

```
> basex
BaseX 9.3.3 RC1 [Standalone]
Try 'help' to get more information.
Mai 12, 2020 10:03:58 AM org.jline.utils.Log logr
WARNUNG: Unable to create a system terminal, creating a dumb terminal (enable debug logging for more information)
> 
```

For BaseX 10 (…no release date available yet), we have planned some upgrades (Jetty 10, Java 11). We could include JLine in the list.

**Comment by dimitarp** [comment]
Thanks for looking at it. I understand your concerns. You can drop the PR for now or merge it in another branch, it's your call :). Maybe you can only cherry pick e550921 since it is a bug fix.

I will continue to regularly rebase the branch on master, because I'm pursuing a different goal: compiling basex to a [native binary executable with GraalVM](https://github.com/dimitarp/basex/tree/basex-native) and I couldn't get it with JLine2 since it has some binary libs in the JAR.

I haven't seen the warning which you posted, but I think I found the reason: https://github.com/jline/jline3#jansi-vs-jna
> On the Windows platform, relying on native calls is mandatory, so you need to have either Jansi or JNA library in your classpath along with the jline-terminal-jansi or jline-terminal-jna jar. Failing to do so will create a dumb terminal with no advanced capabilities.
So basically, beside `jline-reader` and `jline-terminal` a 3rd lib is needed for Windows. I'll add a separate Maven profile which is active only for windows and adds the needed dependency.

**Comment by ChristianGruen** [comment]
> compiling basex to a native binary executable with GraalVM

Exciting; I’m looking forward to your experiences. I think it was @dirkk who was the first to bring up GraalVM and BaseX in our team, but we didn’t pursue it further.

> Maybe you can only cherry pick e550921 since it is a bug fix.

Done. Well, the bug is quite a clear indicator that no one used this Stemmer for ages ;)

> So basically, beside `jline-reader` and `jline-terminal` a 3rd lib is needed for Windows. I'll add a separate Maven profile which is active only for windows and adds the needed dependency.

I see. So I guess the lib needs to be included in the distributions as well. I’ll keep the PR open for some more time.

**Comment by dimitarp** [comment]
> > compiling basex to a native binary executable with GraalVM
> 
> Exciting; I’m looking forward to your experiences. I think it was @dirkk who was the first to bring up GraalVM and BaseX in our team, but we didn’t pursue it further.
Tests are welcome: https://github.com/dimitarp/basex/tree/basex-native/basex-native
I would like to do some performance and functional tests but currently I lack the time to prepare them. Maybe you can give me feedback or provide me with some tests.

**Comment by ChristianGruen** [comment]
…still mentally queued, I hope I’ll make it this week.

**Comment by ChristianGruen** [comment]
@dimitarp I didn’t manage yet to create the binary executable so far:

1. I downloaded and unzipped https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.1.0/graalvm-ce-java11-windows-amd64-20.1.0.zip
2. I jumped to the `bin` directory and installed the native image generator via `gu install native-image`.
3. I adjusted the JDK path for Maven (`set JAVA_HOME=....\graalvm-ce-java11-20.1.0\)
4. I switched to your `basex-native` branch, I compiled basex-core, I ran `mvn --projects basex-native --activate-profiles native clean package`, and I got:

```
[ERROR] Failed to execute goal org.graalvm.nativeimage:native-image-maven-plugin:20.0.0:native-image (basex-cli) on project basex-native: Image building on Java 11+ without native-image requires ```

5. After `set MAVEN_OPTS=--add-exports=java.base/jdk.internal.module=ALL-UNNAMED`, I get the same error.
6.`set MAVEN_OPTS='--add-exports=java.base/jdk.internal.module=ALL-UNNAMED'` doesn’t work on Windows. When running Maven again, I get:

```
Error: Could not find or load main class '--add-exports=java.base.jdk.internal.module=ALL-UNNAMED'
Caused by: java.lang.ClassNotFoundException: '--add-exports=java.base.jdk.internal.module=ALL-UNNAMED'
```

Did you experience a similar behavior?

**Comment by ChristianGruen** [comment]
One of the simplest test cases you could try is: `(1 to 100000000)[. = 0]`.

**Comment by ChristianGruen** [comment]
@dimitarp Did you have a chance to look at the issues I encountered? I guess they are specific to the Windows platform?

**Comment by ChristianGruen** [comment]
We’ll stick with JLine 2 in the middle term; I decided to drop the PR for now. Thanks.
